### ISSUE TITLE
Issue with tags in a nested snippet model

### ISSUE THREAD

**Issue by dejaeghered** [wagtail/wagtail#5414]
### Issue Summary

I have a simple Quiz snippet whith questions and answers.   The answers are nested and in relation to the question.  The answers also have tags configured to them.  On saving this causes an exception because of the nested tags and formsets.
**['ManagementForm data is missing or has been tampered with']**

### Steps to Reproduce

```python
@register_snippet
@python_2_unicode_compatible
class QuizQuestion(ClusterableModel):
    question = models.CharField(max_length=255)

    panels = [
        FieldPanel('question', classname="full"),
        InlinePanel('quiz_answer_list', label="Answers"),
    ]

    def __str__(self):
        return self.vraag


class QuizAnswerTag(TaggedItemBase):
    content_object = ParentalKey('tools.QuizAnswer', on_delete=models.CASCADE, related_name='quiz_tagged_items')


class QuizAnswer(ClusterableModel, Orderable):
    question = ParentalKey(QuizQuestion, related_name='quiz_answer_list')
    answer = models.CharField(max_length=255)
    tags = ClusterTaggableManager(through=QuizAnswerTag, blank=True)

    panels = [
        FieldPanel('answer'),
        FieldPanel('tags'),
    ]
```

Create a question with some answers.  Saving will throw the exception.

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: yes 

### Technical details

Python: 3.6.8
Django 2.1.9
Wagtail version: 2.5.x

**Comment by DanielSwain** [comment]
I have encountered the same problem with a similar setup.  My setup is similar to the above, but I'm not using snippets - just regular Django models.  If I try to delete a `ProductVariation` or even just change the `is_active` field on a `ProductVariation` and then click Save, I get the **['ManagementForm data is missing or has been tampered with']** error.  Here are the relevant parts of my models:

```
class BaseModel(models.Model):
    created_by = models.ForeignKey(User, null=True, blank=True, related_name="%(class)s_created_by", on_delete=models.SET_NULL, editable=False)
    modified_by = models.ForeignKey(User, null=True, blank=True, related_name="%(class)s_modified_by", on_delete=models.SET_NULL, editable=False)
    date_created = models.DateTimeField(auto_now_add=True, editable=False)
    date_modified = models.DateTimeField(auto_now=True, editable=False)

    class Meta:
        abstract = True


class ProductVariation(BaseModel, ClusterableModel):
    product = ParentalKey('Product', on_delete=models.PROTECT, related_name='product_variations', null=True)
    is_active = models.BooleanField(default=True)
    ...

    panels = [
        MultiFieldPanel([
            FieldPanel('name'),
            FieldPanel('description'),
            FieldPanel('is_active'),
            FieldPanel('price'),
        ], heading='Product Variation Main Information'),
        ...
    ]


class Product(BaseModel, ClusterableModel):
    name = models.CharField(max_length=255, blank=True, help_text='Used as title in search results')
    ...

    panels = [
        ...
        InlinePanel('product_variations', label="Product Variations")
    ]
```

I did not confirm that the problem occurs in a fresh Wagtail project, but this site was originally deployed with Wagtail 2.1 in June 2018.  On June 28, 2018 I added `is_active` to `ProductVariation`, so I know everything was working fine at that point.  This is the first time the customer (a small ecommerce website) has tried to delete a product variation or change the active status of a product variation, so if a bug has been introduced, I can only say that it happened since 2.1.  The site is now at Wagtail 2.7.  I WAS able to delete product variations fine in the Django admin, so it does seem like it's something in Wagtail.

**Comment by DanielSwain** [comment]
I'm pretty sure the error I encountered above is coming in Wagtail Model Cluster.  The code in [this commit](https://github.com/wagtail/django-modelcluster/commit/bd61cb067307deaf8237f099c0ae0424239a8674) from January 17, 2019 is where the error appears to have been introduced (specifically, line 374:  `if self.formsets`).  In addition to the above models, I also have:

```
class OrderItem(BaseModel, Orderable):
    order = ParentalKey('Order', null=True, on_delete=models.CASCADE, related_name='order_items')
    product = ParentalKey('Product', null=True, on_delete=models.PROTECT)
    product_variation = ParentalKey('ProductVariation', null=True, blank=True, on_delete=models.PROTECT)
```

It appears that the error results when one `Clusterable` model is `ParentalKey`ed to another `Clusterable` model.  In my case, if an order has been placed for a variation of a product, then when `if self.formsets` is evaluated upon saving a `Product`, any `ProductVariation` that has a related `OrderItem` will yield an `OrderItem` formset and so the code from line 375 on is executed.  (When editing a `Product` and it's associated `ProductVariation`s, no `OrderItem` has been specified in the panels definition above - the `if self.formsets` statement (and the subsequent `for formset in self.formsets.values():`) is somehow seeing an `OrderItem` formset through the `Orderable`/`Clusterable` relation back to `ProductVariation`).  When I remove `Orderable` from `OrderItem` and `Clusterable` from `Order` (along with removing some other `Orderable`/`Clusterable` definitions on other related models, then the Save functionality on `Product` works again without raising the **['ManagementForm data is missing or has been tampered with']** Django error.

**Note:**  The above model definitions ended up this way because, in order to show an `Order` edit screen with `OrderItem` inlines, the `OrderItem` inlines had to be defined as `Orderable` with, of course `Order` then being defined, as `Clusterable`, and other related models also having to be defined as `Orderable`/`Clusterable` as necessary.  I realize that this structure likely needs to change for reasons other than the error that has cropped up.

Hoping that @gasman or @Sam-Costigan can take a look at this issue.

**Comment by johannesvogel** [comment]
We are having the same issue with the site of a customer. Running Wagtail 2.7 on Django 2.2.8, I can reproduce what @DanielSwain is saying. You cannot edit/save any Page, that has a child that is clusterable and (or?) orderable. This is where it happens for us:

```python
class PersonPage(BasePage):
    # ...

    content_panels = [
        InlinePanel('functions', label=_('Functions')),
        # ...
    ]

class Function(index.Indexed, ClusterableModel, Orderable):
    person = ParentalKey(to='cms.PersonPage', related_name='functions', verbose_name=_('Person'))
    # ...
```

Would be really cool if we could get a fix for this, because I currently can't think of a workaround apart from downgrading to Wagtail 2.6 and django-modelcluster 4.3.

**Comment by gasman** [comment]
Hi all - thanks for the reports.

@DanielSwain, if this is indeed a regression in django-modelcluster 4.4 then that's something I'd be particularly keen to investigate - but without a standalone test case to work with (including full model definitions, panel definitions and modeladmin config - assuming modeladmin is the mechanism you're using to set up the forms) it's hard to tell whether your setup is something "within spec" that we'd expect to work. Even if it appeared to work on previous versions of django-modelcluster, it's possible that it was relying on undefined behaviour.

Having multiple `ParentalKey`s on the same model isn't formally supported - in practice I don't _think_ it should be a problem unless any of the parent models are Page models (because then the page model will try to track changes to the child model as part of its own revision history, leading to inconsistencies if the child model receives edits outside the context of that page). The nesting of ClusterableModels (Product -> ProductVariation -> OrderItem) is also something that's raising alarm bells - if you're editing all of those in a single form via nested InlinePanels, then that's exactly the setup that @Sam-Costigan's fix in 4.4 was intended to solve. Conversely, if these aren't all part of the same form, then it may be that one or more of the ParentalKeys should be a ForeignKey instead.

**Comment by johannesvogel** [comment]
Hey @gasman, thank you for your response! I created a simple project where this error occurs:

https://github.com/johannesvogel/wagtail-nested-error

The project contains three models:

- `PersonPage(Page)`
- `Address(index.Indexed, ClusterableModel, Orderable)`
- `AddressTag(TaggedItemBase)`

A person can have zero or more addresses and an address can have zero or more tags. You can save a `PersonPage` if you don't add an address, but once you add an address, it fails with the error described above. If you remove `Address.tags` from the model, it also works. I used tags for the ease of testing it, but it also fails for other models that are connected with a `ParentalKey`.

Hope this helps! Otherwise, just let me know here and I'd be happy to clarify.

**Comment by gasman** [comment]
Thanks @johannesvogel! I can confirm that this works on Wagtail 2.6 with django-modelcluster 4.3 but fails on 4.4, which backs up what @DanielSwain is reporting.

**Comment by gasman** [comment]
It looks like the problem is as follows:

#106 changed child forms to be created as ClusterForm, as a necessary first step towards supporting multi-level nested forms; however, it doesn't provide a way to pass a `formsets` / `exclude_formsets` parameter to the child form to specify which second-level child forms should be included (and if it _did_ provide a way, existing Wagtail releases don't use it). In this case, it will default to adding a formset for every available child relation. (This matches the pre-Django-1.8 behaviour of `forms` / `exclude` on ModelForm, where omitting both of them would give you a form containing all fields.)

Regardless of what the ClusterForm defines, Wagtail will only render the child formsets that have been explicitly listed as InlinePanels. Tag fields are an example of a child relation that _isn't_ managed through a formset / InlinePanel, and as a result, the ClusterForm is expecting a formset that the form submission from Wagtail isn't providing, hence the 'ManagementForm is missing' error. Similarly, in @DanielSwain's case, the ClusterForm is expecting a formset for the ProductVariation -> OrderItem relation that Wagtail doesn't provide.

I think the lowest-friction fix to restore compatibility between the old and new behaviours will be to have ClusterForm silently skip formsets where the management form is missing - this would be released as django-modelcluster 4.5, and probably 5.1 if the eventual 'proper' fix is a breaking change. For the longer term, we really want a mechanism for passing `formsets` explicitly, particularly as I suspect #5126 is also down the lack of parameter-passing - and possibly also make passing `formsets`/`exclude_formsets` compulsory, so that we're finally consistent with Django >=1.8's behaviour for `fields`.

**Comment by gasman** [comment]
PR created: https://github.com/wagtail/django-modelcluster/pull/119

I've made it so that the 'ignore missing formsets' behaviour only applies when formsets are not specified explicitly - this way, when we implement the proper fix of specifying the formsets on the Wagtail side, the 'hackiness' (i.e. ignoring a potentially legitimate error case) is confined to a legacy code path, and so we don't need to consider a backwards-incompatible 6.x release to remove the hack.

**Comment by gasman** [comment]
This fix is now released as django-modelcluster version 4.4.1 and 5.0.1 - the one remaining task we ought to do here is turn @johannesvogel's example project into a test case on the Wagtail side, so that we can catch any future regressions in this setup.
